Documentation/AdditionalFunctionality/PushNotifications.md

```diff
@@ -0,0 +1,133 @@
+    public var token: String
+    public var subscriptions: [String]?
+}
+
+public enum PushNotificationService: String {
+    case apns
+    case fcm
+}
+```
+
+The web service needs a specific route and `Component` to handle push notification registration. The following example shows how to add a `Device` to the currently authenticated `User`.
+
+```swift
+struct RegisterComponent: Component {
+    @Request(\.user) var user: User  // The logged in user
+
+    @Parameter var device: Device  // The `Device` instance in a request body
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Bool> {
+        notificationCenter.register(device, to: user);
+    }
+}
+```
+
+Furthermore, the `NotificationCenter` allows the removal and editing of `Device`s.
+
+## Sending Push Notifications
+
+`Notification`s are structured as follows:
+
+- **alert**: Displayed message on the device which includes title, subtitle, and body.
+- **payload**: Service specific notification settings like sound, badges, images, etc.
+- **data**: Background data.
+
+`Notification`s don't have to include an alert and can just send a silent push notification with background data.
+
+```swift
+public struct Notification {
+    public var alert: Alert
+    public var payload: Payload
+    public var data: Encodable
+}
+```
+
+Example of a `Component` that sends a push notification to an user:
+
+```swift
+struct SendNotification: Component {
+    @Request(\.user) var user: User
+
+    @Environment(\.notificationCenter) var notificationCenter: NotificationCener
+
+    func handle() -> EventLoopFuture<Bool> {
+        notificationCenter.send(alert: .init(title: "Hello There 👋"), to: user)
+    }
+}
+```
+
+The `NotificationCenter` offers the following convenience methods to send push notifications:
+
+- Send to all registered devices
+- Send to a subscription
+- Send to all devices of a user
+- Send to multiple users
+- Send to a specific device
+- Send to a specific token
+
+## Unsolicited Push Notifications
+
+Events can trigger the sending of push notifications. This builds on top of `Component`s and follows the same approach as **[Service-Side Stream Components](../ComponentTypes/ServiceSideStream.md)** with `ObservableObject`s. `Component`s subscribe to `ObservableObject`s that are exposed to the environment. The handle method of the `Component` will be executed every time `@Published` properties of the `ObservableObject` change. `ObservableObject`s also can be scheduled as **[CronJobs](./CronJob.md)**.
+
+The following example shows a _WeatherService_ as a `CronJob` that fetches the weather every day and publishes it to its subscribers using the `@Published` property wrapper. The _AlertComponent_ listens to changes of the _WeatherService_ and sends out push notification based on the current weather. Every `Device` with the **subscription** _weatherSubscription_ will receive this push notification.
+
+```swift
```

Comment:
Please extend this example. It is not quite clear to me, where this `Component` would be placed inside the `Server` declaration. I especially do not understand how "normal" exporters would interact with this `Component`. What part of Apodini would control the lifetime of this `Component` and observe the `weatherService`? What happens to the return-value of the `Component`'s `handle`?

Reply:
Since unsolicited `Component`s are not request based they would be ignored by the exporters in case of APNS. My idea was to either place it inside of the `content` stored property or the `configuration`. It might be confusing to put in `content` with other nested `Component`s like `Group`, which wouldn't really change the lifetime and execution of it since it isn't request based. However, it makes more sense to specify the whole behavior of the server in `content`. Another possibility would be to specify it in `configuration` to clearly separate it from other `Component`s and their control flow. My concept is to create a singleton from a `EventComponent` which is only executed once it is triggered by publishers it listens to.